core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -977,4 +980,35 @@ class KafkaApis(val requestChannel: RequestChannel,
     if (!authorize(request.session, ClusterAction, Resource.ClusterResource))
       throw new ClusterAuthorizationException(s"Request $request is not authorized.")
   }
+
+  def getProtocolVersions(): java.util.List[ProtocolVersion] = {
+    val protocolVersions: java.util.List[ProtocolVersion] = new java.util.ArrayList[ProtocolVersion]()
+    ApiKeys.values().foreach(apiKey => {
+      var index: JShort = JShort.valueOf("-1")
+      val reqs = Protocol.REQUESTS(apiKey.id).flatMap(r => {
+        index = (index + 1).toShort
+        if (r != null)
+          Some(index)
+        else
+          None
+      })
+
+      protocolVersions.add(new ProtocolVersion(apiKey.id, apiKey.name, reqs.toSeq.asJava, Protocol.DEPRECATED_VERSIONS(apiKey.id).toSeq.asJava.asInstanceOf[java.util.List[JShort]]))
+    })
+    protocolVersions
+  }
+
+  def handleProtocolVersionRequest(request: RequestChannel.Request) {
+    val responseHeader = new ResponseHeader(request.header.correlationId)
+    val responseBody = if (!authorize(request.session, Describe, Resource.ClusterResource)) {
+      ProtocolVersionResponse.fromError(Errors.CLUSTER_AUTHORIZATION_FAILED)
+    } else {
+      new ProtocolVersionResponse(Errors.NONE.code(), getProtocolVersions())
+    }
+    requestChannel.sendResponse(new RequestChannel.Response(request, new ResponseSend(request.connectionId, responseHeader, responseBody)))
+  }
+
+  def handleUnknownRequest(request: RequestChannel.Request) {
```

Comment:
Yeah, it will be hard to test. Just want to be sure the server side change from disconnecting is handled well in the clients packaged in Kafka (Java and Scala). It may be something that a systest can help better test. I am okay if its tracked in a separate jira if its difficult to do here.